src/charm.py

```diff
@@ -6,20 +6,355 @@
+
+    @property
+    def _is_traefik_configuration_valid(self) -> bool:
+        if not self._config.is_valid:
+            self._stored.invalid_config = True
+            return False
+
+        self._stored.invalid_config = False
+        return True
+
+    @property
+    def _is_traefik_config_ready(self) -> bool:
+        if not self._is_traefik_configuration_valid:
+            self.unit.status = BlockedStatus("Invalid or missing config. See logs for more info")
+            return False
+
+        # validate ingress-per-unit relation status
+        if not self._ipu_relation:
+            return False
+        if not self.ingress_per_unit.is_ready(self._ipu_relation):
+            return False
+
+        logger.info("Traefik-related configs and relations are ready")
+
+        return True
+
+    def _on_ingress_data_provided(self, event: IngressDataReadyEvent) -> None:
+        if not self._is_traefik_config_ready:
+            logger.info("Traefik-related config or relations not ready. Deferring the event")
+            event.defer()
+            return
+
+        self._generate_traefik_forwardauth_config()
+        self._on_update_status(event)
+
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        if self.access_rules_configured:
+            try:
+                self.validate_access_rules_config(self.model.config.get("access_rules"))
+                self._stored.invalid_config = False
+            except Exception as e:
+                self.unit.status = BlockedStatus("Invalid json configuration, see logs")
+                logger.error(f"Invalid json configuration: {e.args[0]}")
+                self._stored.invalid_config = True
+                return
+
+        self.unit.status = MaintenanceStatus("Configuring the charm")
+
+        if self._is_traefik_config_ready:
+            self._generate_traefik_forwardauth_config()
+
+        self._on_update_status(event)
+
+    def _on_update_status(self, event: EventBase) -> None:
+        """Set the unit status.
+
+        - If the unit is blocked, leave it that way (it means that the config is invalid)
+        - If any mandatory config is missing, the status is blocked
+        - If the ingress-per-unit relation is missing or not ready, the status is blocked
+        - Else status is active
+        """
+        if self._stored.invalid_config is True:
+            return
+        if not self.access_rules_configured:
+            self.unit.status = BlockedStatus("Missing required configuration value: access_rules")
+        # TODO: Non-charms will not require ipu relation, adjust it
+        if not self._ipu_relation:
+            self.unit.status = BlockedStatus("Awaiting to be related via ingress-per-unit")
+            return
+        if not self.ingress_per_unit.is_ready(self._ipu_relation):
+            self.unit.status = BlockedStatus("ingress-per-unit relation is not ready")
+            return
+        else:
```

Comment:
I think `else` is not necessary here if all the condition blocks above can return early.